src/Symfony/Component/String/AbstractString.php

```diff
@@ -0,0 +1,649 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\String;
+
+use Symfony\Component\String\Exception\ExceptionInterface;
+use Symfony\Component\String\Exception\InvalidArgumentException;
+use Symfony\Component\String\Exception\RuntimeException;
+
+/**
+ * Represents a string of abstract characters.
+ *
+ * Unicode defines 3 types of "characters" (bytes, code points and grapheme clusters).
+ * This class is the abstract type to use as a type-hint when the logic you want to
+ * implement doesn't care about the exact variant it deals with.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ * @author Hugo Hamon <hugohamon@neuf.fr>
+ *
+ * @throws ExceptionInterface
+ *
+ * @experimental in 5.0
+ */
+abstract class AbstractString implements \JsonSerializable
+{
+    public const PREG_PATTERN_ORDER = \PREG_PATTERN_ORDER;
+    public const PREG_SET_ORDER = \PREG_SET_ORDER;
+    public const PREG_OFFSET_CAPTURE = \PREG_OFFSET_CAPTURE;
+    public const PREG_UNMATCHED_AS_NULL = \PREG_UNMATCHED_AS_NULL;
+
+    public const PREG_SPLIT = 0;
+    public const PREG_SPLIT_NO_EMPTY = \PREG_SPLIT_NO_EMPTY;
+    public const PREG_SPLIT_DELIM_CAPTURE = \PREG_SPLIT_DELIM_CAPTURE;
+    public const PREG_SPLIT_OFFSET_CAPTURE = \PREG_SPLIT_OFFSET_CAPTURE;
+
+    protected $string = '';
+    protected $ignoreCase = false;
+
+    abstract public function __construct(string $string = '');
+
+    /**
+     * Unwraps instances of AbstractString back to strings.
+     *
+     * @return string[]|array
+     */
+    public static function unwrap(array $values): array
+    {
+        foreach ($values as $k => $v) {
+            if ($v instanceof self) {
+                $values[$k] = $v->__toString();
+            } elseif (\is_array($v) && $values[$k] !== $v = static::unwrap($v)) {
```

Comment:
I do not get the reasoning behind this piece of code.

Is there any gain in writing:
```
if ($values[$k] !== $v) {
   $values[$k] = $v;
}
```

Instead of just:
```
  $values[k] = $v;
```

Reply:
the goal is to avoid writing to the array in case the change would not actually change the array, to avoid triggering a copy of the array.

Reply:
So the PHP engine is not smart enough to figure that out itself, that an assigment to an array should not trigger a copy it is the exact same value?

I did not know that.

Reply:
Correct, see https://bugs.php.net/76150